# keras

In [ ]:
git@github.com:pointOfive/sixty_sixteen.git

In [ ]:
# (1) https://brew.sh
# `/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"`
# (2) https://git-scm.com/download/mac
# `brew install git`
# https://docs.github.com/en/free-pro-team@latest/github/authenticating-to-github/connecting-to-github-with-ssh
# `ssh-keygen -t rsa -b 4096 -C "iam.scottschwartz@gmail.com"`
# `ls -al ~/.ssh`
# pbcopy < ~/.ssh/id_rsa.pub
# <github/settings/SSH and GPG Keys/New SSH key><ctrl-v: paste key><name it><save>

In [1]:
! conda list

# packages in environment at /Users/gck8gd/opt/anaconda3/envs/deep_learning_6018:
#
# Name                    Version                   Build  Channel
appnope                   0.1.0           py38h32f6830_1001    conda-forge
argon2-cffi               20.1.0           py38h4d0b108_1    conda-forge
async_generator           1.10                       py_0    conda-forge
attrs                     20.2.0             pyh9f0ad1d_0    conda-forge
backcall                  0.2.0              pyh9f0ad1d_0    conda-forge
backports                 1.0                        py_2    conda-forge
backports.functools_lru_cache 1.6.1                      py_0    conda-forge
blas                      1.0                         mkl  
bleach                    3.2.1              pyh9f0ad1d_0    conda-forge
brotlipy                  0.7.0           py38h4d0b108_1000    conda-forge
ca-certificates           2020.6.20            hecda079_0    conda-forge
certifi                   2020.6.20        py38h32f

In [2]:
# https://www.tensorflow.org/install
! pip install tensorflow

     |████████████████████████████████| 165.2 MB 577 kB/s eta 0:00:01
     |████████████████████████████████| 3.3 MB 587 kB/s eta 0:00:01
     |████████████████████████████████| 57 kB 492 kB/s eta 0:00:01
     |████████████████████████████████| 42 kB 658 kB/s eta 0:00:01
     |████████████████████████████████| 6.8 MB 681 kB/s eta 0:00:01
     |████████████████████████████████| 127 kB 579 kB/s eta 0:00:01
     |████████████████████████████████| 15.1 MB 680 kB/s eta 0:00:01
     |████████████████████████████████| 65 kB 698 kB/s eta 0:00:01
     |████████████████████████████████| 1.3 MB 589 kB/s eta 0:00:01
     |████████████████████████████████| 3.0 MB 574 kB/s eta 0:00:01
     |████████████████████████████████| 459 kB 593 kB/s eta 0:00:01
     |████████████████████████████████| 114 kB 575 kB/s eta 0:00:01
     |████████████████████████████████| 779 kB 536 kB/s eta 0:00:01
     |████████████████████████████████| 95 kB 698 kB/s eta 0:00:01
     |████████████████████████████████| 298 kB 46

In [3]:
# https://keras.io/getting_started/intro_to_keras_for_engineers/

import tensorflow as tf
from tensorflow.keras import layers


In [4]:
import numpy as np
from scipy import stats

In [6]:
n = 1000
x = stats.uniform(0,10).rvs(n)


In [104]:
# https://mc-stan.org/docs/2_24/functions-reference/lkj-correlation.html
m = [1,0,0]

def mvn(m):
#    cov = np.diag(3*[np.max(m)]) + np.max(m)*np.ones((3,3))
#    return cov #, cov=cov
    return stats.multivariate_normal(mean=m, cov=np.diag(3*[np.max(m)]) + np.max(m)*np.ones((3,3))).rvs(1)
    
mvn(m)

array([-0.30798555, -0.32449238, -0.96098832])

In [105]:
mvn = np.vectorize(mvn)

In [106]:

mvn([[1,0,0],[0,1,0]])

ValueError: cannot reshape array of size 9 into shape (1,1)

In [60]:
[m,0,0],[0,m,0]

([[1, 0], 0, 0], [0, [1, 0], 0])

In [ ]:
inputs = keras.Input(shape=(None, None, 3))

In [ ]:
from tensorflow.keras import layers

# Center-crop images to 150x150
x = CenterCrop(height=150, width=150)(inputs)
# Rescale images to [0, 1]
x = Rescaling(scale=1.0 / 255)(x)

# Apply some convolution and pooling layers
x = layers.Conv2D(filters=32, kernel_size=(3, 3), activation="relu")(x)
x = layers.MaxPooling2D(pool_size=(3, 3))(x)
x = layers.Conv2D(filters=32, kernel_size=(3, 3), activation="relu")(x)
x = layers.MaxPooling2D(pool_size=(3, 3))(x)
x = layers.Conv2D(filters=32, kernel_size=(3, 3), activation="relu")(x)

# Apply global average pooling to get flat feature vectors
x = layers.GlobalAveragePooling2D()(x)

# Add a dense classifier on top
num_classes = 10
outputs = layers.Dense(num_classes, activation="softmax")(x)

In [ ]:
# some Pythoning that we could do
0. affine function
1. little maths: maxout, sigmoid, softplus (not hyperbolic tangent)
2. mixture specifier

or, exceptiallence actuallutely

## Mixture density networks
- this would be a good coding example? 



    
### Output units as parameters for distributions

- parameters of distributions (rather than distributions): $p(y|f(\mathbf{x};\boldsymbol{\theta}))$    
    - gradients of log, multiplication, addition are well-behaved.
    - heterskedastic model: $\sqrt{\frac{f_\beta(x)}{2\pi}} e^{-\frac{f_\beta(x)}{2}(y-f_\mu(x))^2}$ or $\frac{|f_\beta(x)|^{1/2}}{\sqrt{2\pi}^k} e^{-\frac{1}{2}(y-f_\mu(x))^T f_\beta(x)(y-f_\mu(x)) }$  
        - determinant of diagonal matrix $f_\beta(x)$ is product of diagonals
        - if you take the log of this the derivative is $log(f_\beta(x)_i) + (y-f_\mu(x))_if_\beta(x)_i(y-f_\mu(x))_i$
            - log, addition, multiplication...
        - wouldn't want to parameterize by varaince (or you'd be dividing)
        - standard deviation would also be a bad choice... the gradient near 0 can vanish in squaring (making it difficult to learn parameters that are squared). 
    - regardless... covariance matrix must stay PD: eigenvalues of preicison matrix are inverse of those of covariance matrix... so eigenvalues of preicison matrix must be positive.
    - can use softplus to active affine function in positives only
    - for a full covariance matrix $\Sigma(x) = B(x)B(x)^T$ and $B(x)$ may be unconstrained.
        - likelihood calculation here is is $O(n^3)$ -- for determinant and inverse of $\Sigma$, or it's eigen decomposition (or that of $B(x)$)
        
       

Notes I made/took to create my questions


## Key Ideas

- The biggest choices in network architecture involve *depth* and *width*
    - Parameters here are called weights ($W$) and biases ($c$): $g(\mathbf{W}^Tx+\mathbf{c})$
        - You'll most likely be using *ReLU*: $max(0, \mathbf{W}^Tx+\mathbf{c})$ 
            - The caption in Figure 6.3 is very interesting and definitely worth reading!
- Nonlinear layer outputs (i.e., activation functions, $g$ above) chained together allow networks to
    - NOT require manual feature engineering $x_1^2$, $x_1\cdot x_2$, etc.
    - NOT rely on predefined generic transformations $\phi$ into high dimensional spaces
        - while these can capture many nonlinearities, they also *very* likely produce a local smootheness behavior...
    - LEARN transformations $\hat \phi$: $\hat \phi(x; \boldsymbol{\theta})^T\mathbf{w}$
        - that can "do feature engineering for you" with an extremely generic transformation capability if we consider a broad class of $\phi(x; \boldsymbol{\theta})$       
        - and only cost "giving up" on optimizing a convex cost function
        - and in choosing the character of the class of $\phi(x; \boldsymbol{\theta})$, the "manual expertise" of the analyst can still influence the score of transformation considered (by just defining the general functional forms, rather than the specific forms used, which is even better actually)
        
## Optimizing FF Networks

- Nonlinearity (usually) means loss of convexity 
    - so linear solvers and convex optimization procedures can't generally be used for FF networks
- Fitting by the gradient, though, is a universally applicable optimization methodology
    - Other ML methodologies could use it but they usually use shortcuts instead
    - but, when the data is large and those shortcuts are not longer computationally tractable, they'll also just use SGD as the universal optimization method it is
- When optimizing non-convex cost functions, there are no "optimum guarantees" from SGD
    - The gradient is followed: and hopefully it this can drive down the cost to a sufficiently low enough value
        - cross-entropy (negative log likelihood of natural probability distribution) is the usual choice for the cost function
            - this is a result of the principle of maximum likelihood which suggests we use $−\log p(y | x; \theta)$ as our cost function: more on this below...        
            - what is needed from a cost function, though, is that the gradient be a good guide: i.e., if it's pretty flat (near 0) then it doesn't help us...
             - negative log likelihood often protects against this happening (it's a log which counteracts exponentiation that happens, e.g., in softmax, which causes saturation)
                 - you do need to be careful, though, that the log likelihood for many models can become arbitrarily good (larger and larger magnitude negative values)
                     - logistic regression doesn't work like this since 0 is the minimum log likelihood value achievable for this model         
    - Initial values matter: FF networks should initialize weights to small values, biases should be small positive initializations, or 0

        

        
        
## Output units

- *linear* units (*affine functions*) don't pose much challenge for gradient based optimization since they don't saturate
- *logistic sigmoid* function (*inverse logit function*) on *affine functions* (which then define the *logit function*)
    - are generally created by normalizinig *exponentiated affine function*
    - but since $\frac{C \exp(z_0)}{C \exp(z_0) + C \exp(z_1)}$ is constant regardless of $C$, $z_0$ is fixed to $0$ and only $z_1$ is used (or only $z_1$ and $z_2$ in the case of three classes, and so on...)
    - For two classes, then, the normalized sigmoid can be expressed as: $\sigma((2y-1)z) = \frac{1}{1+e^{-(2y-1)z}} = \frac{1}{1+e^{-2yz}e^z} = \frac{e^{yz}}{e^{yz}+e^{-yz}e^z} = \frac{e^{yz}}{e^{yz}+e^{(1-y)z}}$
    - and since this defines the likelihood of a Bernoulli distribution, the cross-entropy (negative log likelihood) is: $-log\left( \frac{1}{1+e^{-(2y-1)z}} \right) = log \left(1+e^{-(2y-1)z} \right) = log \left(1+e^{(1-2y)z} \right) = \zeta((1-2y)z) \underset{y=0,z>0}{\overset{y=1, z<0}{\approx}} |z|$ (whenever $z$ is wrong)
        - which is easily differentiated with respect to $z$, and so, again as with identity units, does not pose much challenge for gradient based optimization since it doesn't saturate (and remains informative) when $z$ is wrong
    - On the other hand, the loss function $(y- \hat p)^2$ gets flatter and flatter the closer $\hat p$ is to $1-y$ (i.e., the more wrong $\hat p$ is)
        - thus we always use cross-entropy (as opposed to, e.g., MSE) for sigmoid outputs
    - Also, note that computing should be done with z where possible, as opposed to $\sigma((2y-1)z)$ which can have under/overflow problems.
    
- *softmax* generalizes the *logistic sigmoid* function to more than two classes
    - It provides a continuous and differentiable, i.e., "soft" version, of argmax
    - Since $\log \text{softmax}(z_i) = z_i - \log \sum_j \exp(z_j) \approx z_i - \underset{j}{\max} z_j$
        - if biggest $z_i$ is correct, then these will mostly cancel, and the likelihood will be driven by other incorrect observations
            - i.e., $z_i$ doesn't matter but $\underset{j\not=i}{\max} z_j$ hurts you so it will try to be small
    - $\log$ cost functions are needed to "reverse the saturation" of softmax under exponentiation
    - $\text{softmax}(\mathbf{z}) = \text{softmax}(\mathbf{z} - \underset{j}{\max} z_j)$ is numerically stable
    - If you don't set $z_0=0$ you are technically "overparameterized"; but, this doesn't really seem to matter in practice [unlike indicator varables where it does for multicollinearity in linear modesl (though not for trees)].

## Hidden units

### ReLU

- Often activation functions have undefined derivatives at certain location (e.g., ReLU)
    - in practice this works fine computationally
        - because we're generally following derivatives to reach a minimum (gradient $\mathbf{0}$) but never actually get there
    - so if the derivatives is not defined at those minimums, it does matter -- they're never actually evaluated
    - not to mention numerical representation is inexact, and left and right derivatives are numerically what's used anyway (and those exist)
- While activation functions are currently an active area of research, ReLU $\max\{0,z\}$ is an *excellent* default choice these days
    - ReLU has no second order derivative (i.e., that's 0), and derivative is 1 when it's activated 
        - so ReLU provide a super easy and predictable gradient surface) 
        - and *most IMPORTANTLY* have a ZERO derivative degregation from saturdation when activated
        - the basic story is models with behavior that's more "linear" are easier to fit with gradient descent.
            -  Internal affine weights are generally initialize to a small positive value (like 0.1) so the activation function starts "on" for most observations (and the derivative exists)

- There are some generalized versions of ReLU 
- Performance of choices beyond ReLU is very hard to predict actual performance 
    - it's a very "trial and error" process
    - and often they just work about the same, only occasionally working better than ReLU...
- These generalizations exist because ReLU is flat -- i.e., can't learn -- when it's inactive -- and the generalizations try to address this

    - $\max\{0,z_i\} + \alpha_i \min\{0,z_i\}$
        - *absolute value rectification*: $\alpha_i=-1$ give $|z|$
        - *leaky ReLU*: $\alpha_i \approx 0.01$
        - *Parametric ReLU (PReLU)*: makes $\alpha_i$ learnable
        - *maxout* units: $f(z)_i = \underset{z_j \in \mathbf{z}[(1+gk−k):gk]}{\max} z_j(=(\mathbf{W}^Tx+\mathbf{b})_j)$
            - $k$ $z_j$ are considered for $f(z)_i$ (one for each $z_j \in \mathbf{z}[(1+gk−k):gk]$), and each $z_j$ is a linear function in $x$ (note that $x$ is a scalar)
                - plotting each of these $k$ $z_j$ versus each $x$ (again, $x$ is just a scalar  here) is a bunch of lines over $x$
                - taking the the supremum of these $k$ linear functions makes a piecewise linear function (with $k$ pieces, and which must be convex)
                - since different lines are defined by weights, and the weights are learned, a convex activation function can be learned... larger $k$ means a more flexible piecewise linear function.
            - often more regularization is required to fit maxout activation functions since the activation functions have $2k$ parameters (defined by the weights and biases) (or $k$ parameter vectors when x is not a scalar and the weights become vectors)
            - and since each $z_j$ is considered in multiple $f(z)_i$, there is some redundancy that projects against (*catastropic forgettting*).
            - there are some ways to reduce the number of necessary weights in maxout activation functions

### Others

- Logistic Sigmoid $\sigma(z) = \frac{1}{1+e^{-z}}$ and Hyperbolic Tangent $tanh(z) = 2\sigma(2z) - 1$
    - saturate everywhere except near 0
    - work as an output layer if there's an appropriate cost function to "reverse" the saturation.
    - $tanh(z)$ resembles liner around 0 (unlike $\sigma(z)$ which is not symmetric around 0) if activations are kept small
        - this linearity can be helpful for gradient calculations 
- Others
    - *identity* can be used to do "factor models" first and feeding factors into a network as opposed to all the data
    - *softmax* can provide an internal discrete probability distribution
    - *RBF* provices template matching
        - but RBF saturates for most $x$... so is hard to fit/optimize good models (i.e., calculate gradients) with RBF...
    - *softplus* does not work as well as ReLU in practice... strangely/counter intuititively...
    - *hard tanh* $\max(−1, min(1, z))$ just looks like a "ReLU'd" version of $tanh$
    



    
## Arichtecture

- A singly hidden layer networks can fit data well
- but deeper (chained) networks (with overall less parameters) tend to provide better generalization(!)
    - Deeper networks encode a prior belief in *composition* of features as relevant, i.e., features of varaition are in turn composed of simpler features of variation
        - and this empirically does seem to have improved generalization properties
- The *universal approximation theorem* says: networks with a "squashing" layer (and this result  shown for ReLU as well) and enough hidden units *can* represent any function (and match it's derivatives)
    - although *actually doing this* is another thing
        - *making* the network give a good approximation/representation of a function means actually getting SGD to find "good enough" network weights (in the context of the given architecture and it's activation functions) while not ending up overfitting...
    - And remember: the *no free lunch* theorem says there's no best algorithm to fit the MLP which will generalize the best universally 
    - And also *notice* that: depending on how complex the function it the number of required hidden units this *"can do it"* can actually become intractible...
    
                
## Back propegation 
- FF networks forward propegate data to a cost function
- Effects of changes to the parameters (gradients) to how incoming x effects the cost function can be found by going the opposite way with "backprop"
- *Backprop* just computes gradients of parameters with respect to the cost function
- Following the gradient (e.g., by SGD) is the *optimzation* procedure
- Gradients are computed using the *chain rule* $\frac{\partial z}{\partial x} = \frac{\partial z}{\partial y}\frac{\partial y}{\partial x}$
    - 6.45/6 show vector generalizations
    - 6.47 shows shows tensor generalizations

## Symbolic Representation of Computational Graphs
- Computational graphs can be formally defined as in figure 6.8
    - these are *symbolic representation*
    - Torch/Caffe take a symbolic representations and inputs and produce *symbol-to-number differentiation*, i.e., the derivatives for these numbers (extended graph is just never exposed)
    - Theano/Tensorflow add nodes to the computational graph which will produce the gradients (e.g., 6.10)
        - higher order derivativs can be calculated as derivatives on the graph with derivatives    
- *Backpropegation* can be defined by a computational graph:
    - This amounts to visiting each edge in a FF network, computing the partial derivative of the next node with respect to the current node; and dot producting that result to the partial derivatives leading into the current node to extend those partial derivatives to the previous nodes (by the *chain rule*)
    - since many expressions depend on the same evaluations, table filling -- called *dynamic programming* -- is used, which will lead to a $O(n^2)$ rather than exponential computational demands 
        - and, actually, given that the usual architecture in a FF network is "chain structure", the computational demand with this appraoch is frequently even less than $O(n^2)$.
    



## From your reading, you should know/be aware of the following:

- The `xor` was used to show that (single layer) perceptrons had limited representational capacity

- Deep networks with only *linear activation functions* can be reduced to a single layer network representing a linear model
    - thus we must use "an affine transformation controlled by learned parameters, followed by a fixed nonlinear function called an activation function"

- Using *calculus of variations*, cost functions can be interpreted as *functionals* (mapping functions to real values)
    - so $f(\textbf{x})$ estimating $E[y|\textbf{x}]$ is found using mean squared error
    - and $f(\textbf{x})$ estimating $median[y|\textbf{x}]$ is found using mean absolute error
    
- We will later see alternative architectures beyond just chained FF networks (at various depths)
    - E.g., CNNs to share parameters, and skip connections to improve gradient flow
    - Parameter reduction though clever re-use can be an important "architecture consideration"
        - e.g., as is done with CNNs and can be done in some inistances with (learned) maxout activation functions

- Backprop is just an algorithm (with recipes given in 6.5.3/4/6/7) that can be computed 
    - in $O(n^2)$ time for $n$ the number of edges in a FF network (and constant time derivatives evaluation)
    - or better (e.g., in settings such as 6.57) when algebraic/mathematical simplifications can be made.
    - This is because it's a general *automatic differentiation* algorithm (a special case of the *reverse mode accumulation* class of algorithms)    
    
- As noted in Section "6.5.8 Complications", there are *many* practical considerations in play when it comes to actual computational implementations of backprop algorithms.  

- Backprop is not the only way to go about computing gradients; but, is generally sufficiently effective within the deep learning community so as to make it the "tool of choice" with a very dominant "market share" in the deep learning context.

- There are some things called *Krylov methods* which allow for computationally feasible approximiate computations involving the Hessian: those are called *Krylov methods*

- The Historical context of FF networks and MLP as narrated in Section 6.6, particularly the takeaway message that FF networks and MLP are the current state of the art solution of the culmination of efforts to produce general and effective function estimation methodology (and are so effective that they are becoming integral tools enabling *other* tools as foundational underlying subcomponents of those methodologies)

- FF networks are function estimation methodologies (Section 6.4.1 Universal Approximation Properties and Depth)
    - in terms of function representation, there are results that show expoenential power in representation as a result of depth. 



